File: server/src/rebasePlanner.ts
// Interactive rebase planner: prepare a plan, then start the rebase by
// injecting an exact todo list through GIT_SEQUENCE_EDITOR.
//
// Ported from `src-tauri/src/rebase_planner.rs` (desktop module).

import fs from "node:fs/promises";
import path from "node:path";
import { AppError, invalidError, unsupportedError } from "./lib/errors.ts";
import { checked, checkedStdout, runGit, type CommandResult } from "./lib/git.ts";
import { repositoryOperationState, type RepositoryOperationState } from "./operationState.ts";

export interface RebasePlanCommit {
  commit: string;
  authorName: string;
  authorEmail: string;
  authoredAt: string;
  subject: string;
  additions: number;
  deletions: number;
  filesChanged: number;
}

export interface RebasePlan {
  target: string;
  targetCommit: string;
  base: string;
  head: string;
  commits: RebasePlanCommit[];
  warnings: string[];
  blockedReason: string | null;
  supportsUpdateRefs: boolean;
}

export interface RebasePlanItem {
  commit: string;
  subject: string;
  action: "pick" | "reword" | "edit" | "squash" | "fixup" | "drop";
  newMessage?: string | null;
}

export interface RebaseStartRequest {
  target: string;
  base: string;
  items: RebasePlanItem[];
  updateRefs: boolean;
}

export interface RebaseStartResult {
  result: CommandResult;
  operation: RepositoryOperationState;
}

async function revParse(inputPath: string, revision: string): Promise<string> {
  const expression = `${revision}^{commit}`;
  return (await checkedStdout(inputPath, ["rev-parse", "--verify", expression])).trim();
}

async function repositoryClean(inputPath: string): Promise<boolean> {
  return (await checkedStdout(inputPath, ["status", "--porcelain=v1", "-z"])).length === 0;
}

async function supportsUpdateRefs(inputPath: string): Promise<boolean> {
  const result = await runGit(inputPath, ["version"]);
  if (result.code !== 0) return false;
  const raw = result.stdout.trim().split(/\s+/).pop() ?? "";
  const [major = "0", minor = "0"] = raw.split(".");
  const majorValue = Number.parseInt(major, 10) || 0;
  const minorValue = Number.parseInt(minor, 10) || 0;
  return majorValue > 2 || (majorValue === 2 && minorValue >= 38);
}

async function gitDir(inputPath: string): Promise<string> {
  return (await checkedStdout(inputPath, ["rev-parse", "--absolute-git-dir"])).trim();
}

async function parseNumstat(
  inputPath: string,
  commit: string,
): Promise<{ additions: number; deletions: number; files: number }> {
  const output = await checkedStdout(inputPath, ["show", "--numstat", "--format=", "--no-renames", commit]);
  let additions = 0;
  let deletions = 0;
  let files = 0;
  for (const line of output.split(/\r?\n/)) {
    if (!line.length) continue;
    const parts = line.split("\t");
    if (parts.length < 3) continue;
    const [add, del] = parts;
    files += 1;
    if (add !== "-") additions += Number.parseInt(add, 10) || 0;
    if (del !== "-") deletions += Number.parseInt(del, 10) || 0;
  }
  return { additions, deletions, files };
}

async function commitRecord(inputPath: string, commit: string): Promise<RebasePlanCommit> {
  const line = (
    await checkedStdout(inputPath, [
      "show",
      "-s",
      "--date=iso-strict",
      "--format=%H%x1f%an%x1f%ae%x1f%aI%x1f%s",
      commit,
    ])
  ).trim();
  const fields = line.split("\u001f");
  if (fields.length !== 5) {
    throw new AppError("command", `unable to parse commit metadata for ${commit}`);
  }
  const { additions, deletions, files } = await parseNumstat(inputPath, commit);
  return {
    commit: fields[0],
    authorName: fields[1],
    authorEmail: fields[2],
    authoredAt: fields[3],
    subject: fields[4],
    additions,
    deletions,
    filesChanged: files,
  };
}

export async function prepareRebasePlan(inputPath: string, target: string): Promise<RebasePlan> {
  if ((await repositoryOperationState(inputPath)).operation !== null) {
    throw invalidError("finish or abort the current Git operation before planning another rebase");
  }
  if (!(await repositoryClean(inputPath))) {
    throw invalidError("interactive rebase requires a clean working tree; commit or stash changes first");
  }

  const targetCommit = await revParse(inputPath, target);
  const head = await revParse(inputPath, "HEAD");
  const base = (await checkedStdout(inputPath, ["merge-base", targetCommit, head])).trim();
  if (!base) throw invalidError("the selected target and HEAD do not have a merge base");

  const range = `${base}..${head}`;
  const merges = await checkedStdout(inputPath, ["rev-list", "--merges", range]);
  const mergeCommits = merges.split(/\r?\n/).filter((line) => line.trim().length > 0).length;
  const blockedReason = mergeCommits > 0
    ? `This branch contains ${mergeCommits} merge commit${mergeCommits === 1 ? "" : "s"}. The first planner slice intentionally refuses to flatten merge topology.`
    : null;

  const list = await checkedStdout(inputPath, ["rev-list", "--reverse", "--topo-order", "--no-merges", range]);
  const commits: RebasePlanCommit[] = [];
  for (const commit of list.split(/\r?\n/).map((line) => line.trim()).filter((line) => line.length > 0)) {
    commits.push(await commitRecord(inputPath, commit));
  }

  const warnings =
    targetCommit === base
      ? []
      : [
          "The selected target is not the current branch's merge base. Chrono will replay the branch commits from the common base onto the selected target.",
        ];

  return {
    target,
    targetCommit,
    base,
    head,
    commits,
    warnings,
    blockedReason,
    supportsUpdateRefs: await supportsUpdateRefs(inputPath),
  };
}

function shellQuote(value: string): string {
  return `'${value.replaceAll("'", "'\\''")}'`;
}

function sanitizeSubject(subject: string): string {
  return subject.replace(/[\r\n]/g, " ");
}

// Find branches (not checked out in any worktree) that point at commits in the
// rewrite set, so the todo can update them with `update-ref`.
async function updateRefsByCommit(
  inputPath: string,
  commits: Set<string>,
): Promise<Map<string, string[]>> {
  const worktrees = await checkedStdout(inputPath, ["worktree", "list", "--porcelain"]);
  const checkedOut = new Set<string>();
  for (const line of worktrees.split(/\r?\n/)) {
    if (line.startsWith("branch ")) checkedOut.add(line.slice("branch ".length).trim());
  }
  const refs = await checkedStdout(inputPath, [
    "for-each-ref",
    "--format=%(refname)%00%(objectname)",
    "refs/heads",
  ]);
  const byCommit = new Map<string, string[]>();
  for (const line of refs.split(/\r?\n/)) {
    if (!line.length) continue;
    const separator = line.indexOf("\0");
    if (separator < 0) continue;
    const reference = line.slice(0, separator);
    const commit = line.slice(separator + 1);
    if (checkedOut.has(reference) || !commits.has(commit)) continue;
    const existing = byCommit.get(commit) ?? [];
    existing.push(reference);
    byCommit.set(commit, existing);
  }
  for (const values of byCommit.values()) values.sort();
  return byCommit;
}

async function buildTodo(
  items: RebasePlanItem[],
  updateRefs: Map<string, string[]>,
): Promise<string> {
  let todo = "";
  let previousKept = false;
  for (const item of items) {
    const action = item.action;
    if (!["pick", "reword", "edit", "squash", "fixup", "drop"].includes(action)) {
      throw invalidError(`unknown rebase action: ${action}`);
    }
    if ((action === "squash" || action === "fixup") && !previousKept) {
      throw invalidError(`${action} cannot be the first surviving action in an interactive rebase`);
    }
    const subject = sanitizeSubject(item.subject);
    if (action === "reword") {
      const message = item.newMessage?.trim();
      if (!message) {
        throw invalidError(`reword for ${item.commit} requires a new commit message`);
      }
      todo += `pick ${item.commit} ${subject}\n`;
      todo += `exec git commit --amend -m ${shellQuote(message)}\n`;
      previousKept = true;
    } else if (action === "drop") {
      todo += `drop ${item.commit} ${subject}\n`;
    } else {
      todo += `${action} ${item.commit} ${subject}\n`;
      previousKept = true;
    }
    const references = updateRefs.get(item.commit);
    if (references) {
      for (const reference of references) todo += `update-ref ${reference}\n`;
    }
  }
  if (!previousKept) {
    throw invalidError(
      "the plan drops every commit; use reset/branch operations for that history rewrite instead",
    );
  }
  return todo;
}

// Windows git ships `sh` (git-bash); on Unix plain `sh` works. Both accept a
// quoted script path as the first argument.
function shellCommand(scriptPath: string): string {
  return process.platform === "win32" ? `sh.exe ${shellQuote(scriptPath)}` : `sh ${shellQuote(scriptPath)}`;
}

export async function startRebasePlan(
  inputPath: string,
  request: RebaseStartRequest,
): Promise<RebaseStartResult> {
  const current = await prepareRebasePlan(inputPath, request.target);
  if (current.blockedReason) throw invalidError(current.blockedReason);
  if (request.base !== current.base) {
    throw invalidError("the rebase base changed since the plan was loaded; refresh the plan before starting");
  }

  const expected = new Set(current.commits.map((item) => item.commit));
  const supplied = new Set(request.items.map((item) => item.commit));
  if (expected.size !== supplied.size || !request.items.every((item) => expected.has(item.commit))) {
    throw invalidError("the submitted rebase plan does not contain exactly the commits from the current plan");
  }

  const updateRefs = new Map<string, string[]>();
  if (request.updateRefs) {
    if (!current.supportsUpdateRefs) {
      throw unsupportedError("this Git version does not support rebase --update-refs");
    }
    const byCommit = await updateRefsByCommit(inputPath, expected);
    for (const [commit, refs] of byCommit) updateRefs.set(commit, refs);
  }

  const todo = await buildTodo(request.items, updateRefs);
  const directory = await gitDir(inputPath);
  const todoPath = path.join(directory, "chrono-rebase-todo");
  const editorPath = path.join(directory, "chrono-sequence-editor.sh");
  await fs.writeFile(todoPath, todo);
  await fs.writeFile(editorPath, '#!/bin/sh\ncat "$CHRONO_REBASE_TODO" > "$1"\n');
  if (process.platform !== "win32") {
    await fs.chmod(editorPath, 0o700);
  }

  const rebaseArgs = ["rebase", "--interactive"];
  if (request.updateRefs) rebaseArgs.push("--update-refs");
  rebaseArgs.push("--onto", current.targetCommit, current.base);

  const editor = process.platform === "win32" ? "cmd /c exit 0" : "true";
  const result = await runGit(inputPath, rebaseArgs, {
    GIT_SEQUENCE_EDITOR: shellCommand(editorPath),
    CHRONO_REBASE_TODO: todoPath,
    GIT_EDITOR: editor,
  });
  const commandResult: CommandResult = {
    stdout: result.stdout,
    stderr: result.stderr,
    exitCode: result.code,
  };
  const operation = await repositoryOperationState(inputPath);

  await fs.rm(todoPath, { force: true });
  await fs.rm(editorPath, { force: true });

  if (commandResult.exitCode !== 0 && operation.operation === null) {
    throw new AppError("command", `git rebase --interactive: ${commandResult.stderr.trim()}`);
  }

  return { result: commandResult, operation };
}

File: src-tauri/src/rebase_planner.rs
use crate::error::{AppError, AppResult};
use crate::models::{CommandResult, RebasePlan, RebasePlanCommit, RebasePlanItem, RebaseStartRequest, RebaseStartResult};
use crate::operation_state;

#[cfg(not(target_os = "android"))]
mod desktop {
    use std::collections::{HashMap, HashSet};
    use std::fs;
    use std::path::{Path, PathBuf};
    use std::process::{Command, Output};

    use super::*;

    fn run_git(path: &str, args: &[&str]) -> AppResult<Output> {
        let mut command = Command::new("git");
        command.arg("-C").arg(path).args(args);
        command.env("GIT_TERMINAL_PROMPT", "0");
        Ok(command.output()?)
    }

    fn command_result(output: Output) -> CommandResult {
        CommandResult {
            stdout: String::from_utf8_lossy(&output.stdout).into_owned(),
            stderr: String::from_utf8_lossy(&output.stderr).into_owned(),
            exit_code: output.status.code().unwrap_or(-1),
        }
    }

    fn checked(path: &str, args: &[&str]) -> AppResult<CommandResult> {
        let result = command_result(run_git(path, args)?);
        if result.exit_code == 0 {
            Ok(result)
        } else {
            Err(AppError::Command(format!("git {}: {}", args.join(" "), result.stderr.trim())))
        }
    }

    fn rev_parse(path: &str, revision: &str) -> AppResult<String> {
        let expression = format!("{revision}^{{commit}}");
        Ok(checked(path, &["rev-parse", "--verify", &expression])?.stdout.trim().to_string())
    }

    fn repository_clean(path: &str) -> AppResult<bool> {
        Ok(checked(path, &["status", "--porcelain=v1", "-z"])?.stdout.is_empty())
    }

    fn supports_update_refs(path: &str) -> bool {
        let Ok(version) = checked(path, &["version"]) else { return false };
        let raw = version.stdout.split_whitespace().last().unwrap_or("");
        let mut parts = raw.split('.');
        let major = parts.next().and_then(|value| value.parse::<u32>().ok()).unwrap_or(0);
        let minor = parts.next().and_then(|value| value.parse::<u32>().ok()).unwrap_or(0);
        major > 2 || (major == 2 && minor >= 38)
    }

    fn git_dir(path: &str) -> AppResult<PathBuf> {
        Ok(PathBuf::from(checked(path, &["rev-parse", "--absolute-git-dir"])?.stdout.trim()))
    }

    fn parse_numstat(path: &str, commit: &str) -> AppResult<(usize, usize, usize)> {
        let output = checked(path, &["show", "--numstat", "--format=", "--no-renames", commit])?.stdout;
        let mut additions = 0usize;
        let mut deletions = 0usize;
        let mut files = 0usize;
        for line in output.lines() {
            let mut fields = line.splitn(3, '\t');
            let Some(add) = fields.next() else { continue };
            let Some(del) = fields.next() else { continue };
            if fields.next().is_none() { continue; }
            files += 1;
            if add != "-" { additions = additions.saturating_add(add.parse().unwrap_or(0)); }
            if del != "-" { deletions = deletions.saturating_add(del.parse().unwrap_or(0)); }
        }
        Ok((additions, deletions, files))
    }

    fn commit_record(path: &str, commit: &str) -> AppResult<RebasePlanCommit> {
        let line = checked(path, &["show", "-s", "--date=iso-strict", "--format=%H%x1f%an%x1f%ae%x1f%aI%x1f%s", commit])?.stdout;
        let fields: Vec<&str> = line.trim().split('\u{1f}').collect();
        if fields.len() != 5 {
            return Err(AppError::Command(format!("unable to parse commit metadata for {commit}")));
        }
        let (additions, deletions, files_changed) = parse_numstat(path, commit)?;
        Ok(RebasePlanCommit {
            commit: fields[0].to_string(),
            author_name: fields[1].to_string(),
            author_email: fields[2].to_string(),
            authored_at: fields[3].to_string(),
            subject: fields[4].to_string(),
            additions,
            deletions,
            files_changed,
        })
    }

    pub fn prepare_rebase_plan(path: &str, target: &str) -> AppResult<RebasePlan> {
        if operation_state::repository_operation_state(path)?.operation.is_some() {
            return Err(AppError::Invalid("finish or abort the current Git operation before planning another rebase".into()));
        }
        if !repository_clean(path)? {
            return Err(AppError::Invalid("interactive rebase requires a clean working tree; commit or stash changes first".into()));
        }

        let target_commit = rev_parse(path, target)?;
        let head = rev_parse(path, "HEAD")?;
        let base = checked(path, &["merge-base", &target_commit, &head])?.stdout.trim().to_string();
        if base.is_empty() {
            return Err(AppError::Invalid("the selected target and HEAD do not have a merge base".into()));
        }

        let range = format!("{base}..{head}");
        let merge_commits = checked(path, &["rev-list", "--merges", &range])?.stdout
            .lines().filter(|line| !line.trim().is_empty()).count();
        let blocked_reason = if merge_commits > 0 {
            Some(format!("This branch contains {merge_commits} merge commit{}. The first planner slice intentionally refuses to flatten merge topology.", if merge_commits == 1 { "" } else { "s" }))
        } else { None };

        let list = checked(path, &["rev-list", "--reverse", "--topo-order", "--no-merges", &range])?.stdout;
        let mut commits = Vec::new();
        for commit in list.lines().map(str::trim).filter(|line| !line.is_empty()) {
            commits.push(commit_record(path, commit)?);
        }

        let warnings = if target_commit == base {
            Vec::new()
        } else {
            vec!["The selected target is not the current branch's merge base. GitAhead will replay the branch commits from the common base onto the selected target.".into()]
        };

        Ok(RebasePlan {
            target: target.to_string(),
            target_commit,
            base,
            head,
            commits,
            warnings,
            blocked_reason,
            supports_update_refs: supports_update_refs(path),
        })
    }

    fn shell_quote(value: &str) -> String {
        format!("'{}'", value.replace('\'', "'\\''"))
    }

    fn sanitize_subject(subject: &str) -> String {
        subject.replace(['\r', '\n'], " ")
    }

    fn update_refs_by_commit(path: &str, commits: &HashSet<String>) -> AppResult<HashMap<String, Vec<String>>> {
        let checked_out: HashSet<String> = checked(path, &["worktree", "list", "--porcelain"])?.stdout
            .lines()
            .filter_map(|line| line.strip_prefix("branch "))
            .map(str::to_string)
            .collect();
        let refs = checked(path, &["for-each-ref", "--format=%(refname)%00%(objectname)", "refs/heads"])?.stdout;
        let mut by_commit: HashMap<String, Vec<String>> = HashMap::new();
        for line in refs.lines() {
            let Some((reference, commit)) = line.split_once('\0') else { continue };
            if checked_out.contains(reference) || !commits.contains(commit) { continue; }
            by_commit.entry(commit.to_string()).or_default().push(reference.to_string());
        }
        for refs in by_commit.values_mut() { refs.sort(); }
        Ok(by_commit)
    }

    fn build_todo(items: &[RebasePlanItem], update_refs: &HashMap<String, Vec<String>>) -> AppResult<String> {
        let mut todo = String::new();
        let mut previous_kept = false;
        for item in items {
            let action = item.action.as_str();
            if !matches!(action, "pick" | "reword" | "edit" | "squash" | "fixup" | "drop") {
                return Err(AppError::Invalid(format!("unknown rebase action: {action}")));
            }
            if matches!(action, "squash" | "fixup") && !previous_kept {
                return Err(AppError::Invalid(format!("{} cannot be the first surviving action in an interactive rebase", action)));
            }
            let subject = sanitize_subject(&item.subject);
            match action {
                "reword" => {
                    let message = item.new_message.as_deref().map(str::trim).filter(|value| !value.is_empty())
                        .ok_or_else(|| AppError::Invalid(format!("reword for {} requires a new commit message", item.commit)))?;
                    todo.push_str(&format!("pick {} {}\n", item.commit, subject));
                    todo.push_str(&format!("exec git commit --amend -m {}\n", shell_quote(message)));
                    previous_kept = true;
                }
                "drop" => {
                    todo.push_str(&format!("drop {} {}\n", item.commit, subject));
                }
                _ => {
                    todo.push_str(&format!("{} {} {}\n", action, item.commit, subject));
                    previous_kept = true;
                }
            }
            if let Some(references) = update_refs.get(&item.commit) {
                for reference in references { todo.push_str(&format!("update-ref {}\n", reference)); }
            }
        }
        if !previous_kept {
            return Err(AppError::Invalid("the plan drops every commit; use reset/branch operations for that history rewrite instead".into()));
        }
        Ok(todo)
    }

    fn normalized_script_path(path: &Path) -> String {
        path.to_string_lossy().replace('\\', "/")
    }

    pub fn start_rebase_plan(path: &str, request: &RebaseStartRequest) -> AppResult<RebaseStartResult> {
        let current = prepare_rebase_plan(path, &request.target)?;
        if let Some(reason) = current.blocked_reason.as_ref() {
            return Err(AppError::Invalid(reason.clone()));
        }
        if request.base != current.base {
            return Err(AppError::Invalid("the rebase base changed since the plan was loaded; refresh the plan before starting".into()));
        }

        let expected: HashSet<String> = current.commits.iter().map(|item| item.commit.clone()).collect();
        let supplied: HashSet<String> = request.items.iter().map(|item| item.commit.clone()).collect();
        if expected != supplied || supplied.len() != request.items.len() {
            return Err(AppError::Invalid("the submitted rebase plan does not contain exactly the commits from the current plan".into()));
        }

        let update_refs = if request.update_refs {
            if !current.supports_update_refs {
                return Err(AppError::Unsupported("this Git version does not support rebase --update-refs".into()));
            }
            update_refs_by_commit(path, &expected)?
        } else { HashMap::new() };
        let todo = build_todo(&request.items, &update_refs)?;
        let git_dir = git_dir(path)?;
        let todo_path = git_dir.join("gitahead-rebase-todo");
        let editor_path = git_dir.join("gitahead-sequence-editor.sh");
        fs::write(&todo_path, todo)?;
        fs::write(&editor_path, "#!/bin/sh\ncat \"$GITAHEAD_REBASE_TODO\" > \"$1\"\n")?;
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let mut permissions = fs::metadata(&editor_path)?.permissions();
            permissions.set_mode(0o700);
            fs::set_permissions(&editor_path, permissions)?;
        }

        let sequence_editor = format!("sh {}", shell_quote(&normalized_script_path(&editor_path)));
        let mut command = Command::new("git");
        command.arg("-C").arg(path)
            .arg("rebase").arg("--interactive");
        if request.update_refs { command.arg("--update-refs"); }
        command.arg("--onto").arg(&current.target_commit)
            .arg(&current.base);
        command.env("GIT_TERMINAL_PROMPT", "0");
        command.env("GIT_SEQUENCE_EDITOR", sequence_editor);
        command.env("GITAHEAD_REBASE_TODO", normalized_script_path(&todo_path));
        command.env("GIT_EDITOR", "true");
        let output = command.output()?;
        let result = command_result(output);
        let state = operation_state::repository_operation_state(path)?;

        let _ = fs::remove_file(&todo_path);
        let _ = fs::remove_file(&editor_path);

        if result.exit_code != 0 && state.operation.is_none() {
            return Err(AppError::Command(format!("git rebase --interactive: {}", result.stderr.trim())));
        }

        Ok(RebaseStartResult { result, operation: state })
    }
}

#[cfg(target_os = "android")]
mod android {
    use super::*;

    pub fn prepare_rebase_plan(_path: &str, _target: &str) -> AppResult<RebasePlan> {
        Err(AppError::Unsupported("interactive rebase planning is desktop-only in this slice".into()))
    }

    pub fn start_rebase_plan(_path: &str, _request: &RebaseStartRequest) -> AppResult<RebaseStartResult> {
        Err(AppError::Unsupported("interactive rebase execution is desktop-only in this slice".into()))
    }
}

#[cfg(not(target_os = "android"))]
pub use desktop::{prepare_rebase_plan, start_rebase_plan};
#[cfg(target_os = "android")]
pub use android::{prepare_rebase_plan, start_rebase_plan};
